fish-rust/src/builtins/string/shorten.rs

```diff
@@ -0,0 +1,244 @@
+        wopt(L!("max"), required_argument, 'm'),
+        wopt(L!("no-newline"), no_argument, 'N'),
+        wopt(L!("left"), no_argument, 'l'),
+        wopt(L!("quiet"), no_argument, 'q'),
+    ];
+    const SHORT_OPTIONS: &'static wstr = L!(":c:m:Nlq");
+
+    fn parse_opt(&mut self, optarg: Option<&'args wstr>, c: char) -> Result<(), StringError> {
+        match c {
+            'c' => {
+                self.ellipsis = optarg.unwrap();
+                self.ellipsis_width = width_without_escapes(self.ellipsis, 0);
+            }
+            'm' => self.max = Some(fish_wcstol(optarg.unwrap())?.try_into()?),
+            'N' => self.no_newline = true,
+            'l' => self.shorten_from = Direction::Left,
+            'q' => self.quiet = true,
+            _ => return Err(StringError::UnknownOption),
+        }
+        return Ok(());
+    }
+
+    fn handle(
+        &mut self,
+        _parser: &mut parser_t,
+        streams: &mut io_streams_t,
+        optind: &mut usize,
+        args: &[&wstr],
+    ) -> Option<libc::c_int> {
+        let mut min_width = usize::MAX;
+        let mut inputs = Vec::new();
+
+        let iter = Arguments::new(args, optind, streams, true);
+
+        if Some(0) == self.max {
+            // Special case: Max of 0 means no shortening.
+            // This makes this more reusable, so you don't need special-cases like
+            //
+            // if test $shorten -gt 0
+            //     string shorten -m $shorten whatever
+            // else
+            //     echo whatever
+            // end
+            for (arg, _) in iter {
+                streams.out.append(arg);
+                streams.out.append1('\n');
+            }
+            return STATUS_CMD_OK;
+        }
+
+        for (arg, _) in iter {
+            // Visible width only makes sense line-wise.
+            // So either we have no-newlines (which means we shorten on the first newline),
+            // or we handle the lines separately.
+            let mut splits = split_string(&arg, '\n').into_iter();
+            if self.no_newline && splits.len() > 1 {
+                let mut s = match self.shorten_from {
+                    Direction::Right => splits.next(),
+                    Direction::Left => splits.last(),
+                }
+                .unwrap();
+                s.push_utfstr(self.ellipsis);
+                let width = width_without_escapes(&s, 0);
+
+                if width > 0 && width < min_width {
+                    min_width = width;
+                }
+                inputs.push(s);
+            } else {
+                for s in splits {
+                    let width = width_without_escapes(&s, 0);
+                    if width > 0 && width < min_width {
+                        min_width = width;
+                    }
+                    inputs.push(s);
+                }
+            }
+        }
+
+        let ourmax: usize = self.max.unwrap_or(min_width);
```

Comment:
> nit regarding the earlier // TODO: Can we have negative width comment.
If easily possible, I'd have preferred putting the behavior fix commit first, so that such questions never arise.

One issue here is that the earlier Rust code before the fix panicked with a `capacity overflow`, and backporting to C++ so you have to rebase with a C++ bugfix is weird, you kinda need to squash the fix with the port, which is not ideal. I don't like this situation. In this case it was easier to fix the bug than port the bug

Reply:
> and backporting to C++ so you have to rebase with a C++ bugfix is weird, so you kinda need to squash the fix with the port, which is not ideal. I don't like this situation. In this case it was easier to fix the bug than port the bug

one recipe is to rebase, breaking before the rewrite to add a commit that fixes the bug in C++. Then revert that commit and continue the rebase. There are no conflicts because of the revert. Of course the bug fix needs to be applied again on the Rust side (before or after the rebase), and squashed (along with the revert).

Reply:
I think the conclusion here is to squash this all into one commit, I don't think this needs to be backported for a 3.6.2 release, as fixing the C++ code is not as trivial as #9900. If we want to backport this please inform me and make a PR towards master so I can rebase on that when I rebase for #9900

Reply:
any order is fine here (hence nit) but in general it's better to put "smaller" commits first because the larger ones are more likely to be changed further.
Squashing the behavior change into the port is not something we should do (but there is no need to right)

Reply:
Just to make it clear: I do not want to write a C++ version of the behavior fix. The initial porting-commit still has bugged behavior (but different, as in crashing!). 

My understanding is that we do not want to backport the fix, which will leave the "broken" commit in place if we do not squash it. But as you are telling me, that is what we want? If so please mark this as resolved.

> I'd have preferred putting the behavior fix commit first, so that such questions never arise.

I agree that we should ideally have the fix separately, but without writing a C++ version of the fix there is very little I can do to improve the commit ordering.